reference: add user documentation for auto_random feature (#1875)
pingcap/docs: reference/sql/attributes/auto-random.md

+ If the `INSERT` statement explicitly specifies the values of the integer primary key column, TiDB saves these values as they are, which works similarly to the `AUTO_INCREMENT` attribute.

TiDB automatically assigns values in the following way:

The highest five digits of the row value in binary (namely, shard bits) is determined by the starting time of the current transaction. The remaining digits are assigned values in an auto-increment way.

To use different number of shard bits, append a pair of parentheses to `AUTO_RANDOM` and specify the desired number of shard bits in the parentheses. See the following example:

{{< copyable "sql" >}}

```sql
create table t (a int primary key auto_random(3), b varchar(255))
```

In the above `CREATE TABLE` statement, `3` shard bits are specified. The range of the number of shard bits is `[1, field_max_bits)`. `field_max_bits` is the length of bits occupied by the primary key column.

> In the `TIDB_ROW_ID_SHARDING_INFO` row of the `information_schema.tables` system table, the value of the table with the `AUTO_RANDOM` attribute is `PK_AUTO_RANDOM_BITS=x`. In this value, `x` is the number of shard bits.

Reviewer: ```suggestion
For tables with the `AUTO_RANDOM` attribute, the value of the `TIDB_ROW_ID_SHARDING_INFO` column in the `information_schema.tables` system table is `PK_AUTO_RANDOM_BITS=x`. `x` is the number of shard bits.
```
Please confirm with the author about the original Chinese:
对于含有 AUTO_RANDOM 属性的表，在系统表 information_schema.tables 的 TIDB_ROW_ID_SHARDING_INFO 列的值为。。。

Is the above understanding right? @TomShawn

Author: Yes, this is correct.

## Compatibility

TiDB supports parsing the version comment syntax. See the following example:

{{< copyable "sql" >}}

```sql
create table t (a int primary key /*T!30100 auto_random */)
```

{{< copyable "sql" >}}

```sql
create table t (a int primary key auto_random)
```

The above two statements have the same meaning.

> In the result of `show create table`, the `AUTO_RANDOM` attribute is commented out. This comment includes a version number (for example, `/*T!30100 auto_random */`). `30100` in the example indicates that the `AUTO_RANDOM` attribute is introduced in v3.1.0. TiDB of lower versions ignore the `AUTO_RANDOM` attribute in the above comment.

Reviewer: ```suggestion
In the result of `show create table`, the `AUTO_RANDOM` attribute is commented out. This comment includes a version number (for example, `/*T!30100 auto_random */`). Here `30100` indicates that the `AUTO_RANDOM` attribute is introduced in v3.1.0. TiDB of a lower version ignores the `AUTO_RANDOM` attribute in the above comment.
```

> This attribute supports forward compatibility, namely, downgrade compatibility. TiDB earlier than v3.1.0 ignores the `AUTO_RANDOM` attribute of a table (with the above comment), so the table can be used by TiDB of earlier versions.

Reviewer: ```suggestion
This attribute supports forward compatibility, namely, downgrade compatibility. TiDB earlier than v3.1.0 ignores the `AUTO_RANDOM` attribute of a table (with the above comment), so TiDB of earlier versions can also use the table with the attribute.
```

## Restrictions

Pay attention to the following restrictions when you use `AUTO_RANDOM`:

- Specify this attribute for the primary key column **ONLY** of integer type. Otherwise, an error might occur. Refer to [Notes for `alter-primary-key`](#notes-for-alter-primary-key) for exception.
- You cannot use `ALTER TABLE` to modify the `AUTO_RANDOM` attribute, including adding or removing this attribute.
- You cannot change the column type of the primary key column that is specified with `AUTO_RANDOM` attribute.
- You cannot specify `AUTO_RANDOM` and `AUTO_INCREMENT` for the same column at the same time.
- You cannot specify `AUTO_RANDOM` and `DEFAULT` (the default value of a column) for the same column at the same time.

> - It is recommended that you do not explicitly specify a value for the row with the `AUTO_RANDOM` attribute when you insert data. Otherwise, the numeral values to be automatically assigned might be used up in advance.

Reviewer: ```suggestion
- It is **not** recommended that you explicitly specify a value for the column with the `AUTO_RANDOM` attribute when you insert data. Otherwise, the numeral values that can be automatically assigned for this table might be used up in advance.
```
